command/jsonplan/plan.go

```diff
@@ -62,8 +61,9 @@ type change struct {
 	// or "after" is unset (respectively). For ["no-op"], the before and after
 	// values are identical. The "after" value will be incomplete if there are
 	// values within it that won't be known until after apply.
-	Before json.RawMessage `json:"before,omitempty"`
-	After  json.RawMessage `json:"after,omitempty"`
+	Before       json.RawMessage `json:"before,omitempty"`
+	After        json.RawMessage `json:"after,omitempty"`
+	AfterUnknown bool            `json:"after_unknown"`
```

Comment:
Unfortunately I think this is not enough detail for all cases. While a single boolean would be fine for an output of a primitive type, if the value is a collection or structural type then we'll need to replicate the same structure as of `after` but replace all of the unknown values with `true`, like this:

```json
{
  "actions": "create",
  "after": {
    "keepers": {},
    "length": 2,
    "prefix": "",
    "separator": ""
  },
  "after_unknown": {
    "id": true
  }
}
```

I _think_ it might be possible to generate the necessary data structure with [the `Value.Transform` method](https://godoc.org/github.com/zclconf/go-cty/cty#Transform) in `cty`:

```go
    afterUnknown := cty.Transform(newVal, func(path cty.Path, val cty.Value) (cty.Value, error) {
        if !val.Type().IsPrimitiveType() {
            return val, nil // just pass through non-primitives; they already contain our transform results
        }

        if val.IsKnown() {
            // null rather than false here so that known values
            // don't appear at all in JSON serialization of our result
            return cty.NullVal(cty.Bool), nil
        }

        return cty.True, nil
    })
```

(I didn't actually compile/run the above, so it may have some errors in it.)

JSON-marshaling the result of the above (as `ctyjson.Marshal(afterUnknown, afterUnknown.Type())`) should, I think, produce a result good enough for callers to identify specifically what in the resulting structure is unknown.

Reply:
Yeah, this is pretty much exactly what I was thinking - Sentinel will need a way to check that a specific value within the resource is unknown, versus any non-specific attribute within the resource.

Reply:
This is closer, but still not quite right. I wanted your input before I tried to fix things (I'm copy-pasta-ing a bit).

Here's the new output:

```
    {
      "address": "module.all_the_pets.random_pet.module_pets[0]",
      "module_address": "module.all_the_pets",
      "mode": "managed",
      "type": "random_pet",
      "name": "module_pets",
      "index": 0,
      "deposed": true,
      "change": {
        "actions": [
          "Create"
        ],
        "after_unknown": {
          "id": true,
          "keepers": null,
          "length": null,
          "prefix": null,
          "separator": null
        }
      }
    },
``` 

I have a few ideas for iterating over the after_unknown structure to remove `null` values, but figured there was a good chance I was misunderstanding your suggestion in the first place.

Reply:
I think for the `unknown` keys that a simple map with a value of `true` being the only valid value is fine. This could probably be handled during the marshaling process.

Reply:
But yes, this looks like it's on the right track!

Reply:
I'd forgotten that we _do_ include explicit `null` in the JSON serialization in order to retain the shape of object types properly, so it didn't quite behave the way I expected but I think this is fine for a first pass and we can see about filtering out the nulls in a later change as polish.

Constraining this to just include the `true` properties will require some more complexity in the `Transform` function, which will need to be written pretty carefully to avoid creating something that cty considers invalid, so I don't think we should block finishing up the rest of this just to get this perfect, since this now contains the information we need even if it's a little more verbose than it needs to be.

Reply:
Yeah just to be clear, I think in Sentinel we will be looking for the `true` case and not really caring too much about anything else. Basically we want to just ensure that it's in the set when looking for unknown data.

Reply:
Fixed it (with the understanding that this was more for my own human eyeballs then for sentinel)!
Thanks for the feedback @vancluever 

```
    {
      "address": "module.all_the_pets.random_pet.module_pets[0]",
      "module_address": "module.all_the_pets",
      "mode": "managed",
      "type": "random_pet",
      "name": "module_pets",
      "index": 0,
      "deposed": true,
      "change": {
        "actions": [
          "Create"
        ],
        "after_unknown": {
          "id": true,
          "keepers": false,
          "length": false,
          "prefix": false,
          "separator": false
        }
      }
    },```

Reply:
For sure! And yeah, the eyeballs matter here, which is why I think that ultimately `after_unknown` should only include the data that ultimately needs to be there (it could be argued even that boolean value is even not necessary, but I think having a terminating value is a good idea so that object structure can be preserved). 👍 

Ultimately I was thinking something like:

```
{
  ...
  "after_unknown": {
    "id": true,
    "nested": {
      "foo": true
    }
  }
}
```

Reply:
We're gonna cull this down to just the `true` stuff eventually, but gonna leave this as-is for the moment so we can get the rest of this fleshed out and unblock the downstream work depending on this. Of course this means that consumers must rely only on the stuff that is `true` here and assume that the other stuff will go away later, and hopefully we'll get that done before 0.12.0 final to make this nicer.